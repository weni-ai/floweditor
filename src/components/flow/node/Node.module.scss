@import 'variables.module.scss';
@import '~@weni/unnnic-system/src/assets/scss/unnnic.scss';

@keyframes pulse {
  0% {
    transform: scale(1);
  }

  50% {
    transform: scale(1.3) rotate(2deg);
  }

  100% {
    transform: scale(1);
  }
}

.title {
  border-top-right-radius: $unnnic-border-radius-md;
  border-top-left-radius: $unnnic-border-radius-md;
}

.node_container {
  display: flex;

  padding: $node_padding + px;
  z-index: $z_nodes;
  position: relative;

  .uuid {
    user-select: text;
    position: absolute;
    z-index: 10000;
    white-space: nowrap;
    background: #ffffffdd;
    padding: 3px;
    border-radius: 3px;
    cursor: pointer;
    top: 17px;
    align-content: center;
    font-size: 8px;
    font-family: monospace;
    left: 16px;

    &:hover {
      background: #ffffffff;
    }
  }

  &.ghost {
    opacity: 0.5;
    transition: none;
    z-index: $z_dragging;
  }

  &.nondragged {
    opacity: 0.6;
    transition: opacity 400ms ease;
  }

  .node {
    box-shadow: $unnnic-shadow-level-near;

    &:hover {
      outline: $unnnic-color-brand-weni-soft solid $unnnic-border-width-thin;
    }
  }

  &.selected {
    .node {
      outline: $unnnic-color-brand-weni-soft solid $unnnic-border-width-thin;
    }

  }

  &.immutable {
    .node {
      cursor: default;
    }
  }

  .node {
    border-radius: $unnnic-border-radius-md;
    padding: 0;
    user-select: none;
    cursor: pointer;
    min-width: $node_min_width + px;

    flex-shrink: 0;
    transition: transform 200ms;
    font-size: $unnnic-font-size-body-md;
    line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;

    .save_as {
      display: inline;
    }

    .save_result {
      background: #fff;
      padding: $unnnic-spacing-inset-xs;
      text-align: center;

      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-md;
      line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
      color: $unnnic-color-neutral-dark;

      &.border_top {
        border-top: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;
      }

      .result_name {
        font-weight: $unnnic-font-weight-bold;
        display: inline;
      }
    }

    .selected {
      background: #EEFFFC;
    }

    .cropped {
      border: 0px solid green;
      border-top-left-radius: $curvature - 1px;
      border-top-right-radius: $curvature - 1px;
      position: relative;
    }

    .exit_table {
      border: 0px solid red;

      .exits {
        box-shadow: 0px -2px 0px -1px $borders;
        display: flex;
        flex-direction: row;
        flex-wrap: nowrap;
      }
    }

    .add {
      position: absolute;
      right: (-$unnnic-font-size / 2) - $node_padding / 2;
      bottom: (-$unnnic-font-size / 2) - $node_padding / 2;
      visibility: hidden;
      text-decoration: none;
      border-radius: $unnnic-border-radius-sm;
      background-color: $unnnic-color-background-sky;
      outline: $unnnic-border-width-thinner solid $unnnic-color-neutral-clean;
      outline-offset: -$unnnic-border-width-thinner;
      width: 24px;
      height: 24px;
      display: flex;
      justify-content: center;

      span {
        margin-top: 1px;
      }

      &:hover {
        outline-width: 0;
      }
    }

    &:hover {
      .add {
        visibility: visible;
      }

      .drag_group {
        visibility: visible;
      }
    }
  }

  &.flow_start {
    .flow_start_message {
      position: absolute;
      width: 100%;
      z-index: 10000;
      text-transform: uppercase;
      text-align: center;
      top: - $unnnic-font-size-body-md - $unnnic-line-height-md - $unnnic-spacing-stack-xs;
    }
  }
}

.active {
  position: absolute;
  background: #C4EAF5;
  border-radius: 4px;
  font-family: $unnnic-font-family-secondary;
  color: $unnnic-color-neutral-dark;
  display: flex;
  justify-content: center;
  align-items: center;
  font-weight: 500;
  top: -10px;
  left: -10px;
  font-size: 12px;
  width: 22px;
  height: 28px;
  cursor: pointer;
}