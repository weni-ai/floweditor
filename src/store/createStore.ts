import { applyMiddleware, createStore, Middleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import rootReducer from 'store/rootReducer';
import AppState, { initialState } from 'store/state';

const middlewares: Middleware[] = [thunk];

export default (state: AppState = initialState) => {
  const store = createStore(
    rootReducer,
    state,
    composeWithDevTools(applyMiddleware(...middlewares)),
  );

  return store;
};
