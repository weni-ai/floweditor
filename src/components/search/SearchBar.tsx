import * as React from 'react';
import styles from './Search.module.scss';
// @ts-ignore
import { unnnicIcon } from '@weni/unnnic-system';
import { applyVueInReact } from 'vuereact-combined';
import AppState from 'store/state';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  DispatchWithState,
  HandleSearchChange,
  handleSearchChange,
  onUpdateCanvasPositions
} from 'store/thunks';
import { RenderNodeMap, Search } from 'store/flowContext';
import i18n from 'config/i18n';
import Button, { ButtonTypes } from 'components/button/Button';
import TextInputElement from 'components/form/textinput/TextInputElement';

const UnnnicIcon = applyVueInReact(unnnicIcon, {
  vue: {
    componentWrap: 'div',
    slotWrap: 'div',
    componentWrapAttrs: {
      style: {
        all: '',
        display: 'inline-block'
      }
    }
  }
});
export interface SearchStoreProps {
  search?: Search;
  nodes?: RenderNodeMap;
  handleSearchChange?: HandleSearchChange;
}

export class SearchBar extends React.PureComponent<SearchStoreProps, {}> {
  private handleInput(value: string) {
    const nodes = this.findNodes(value);
    this.props.handleSearchChange({
      active: true,
      value: value,
      nodes: nodes,
      selected: 0
    });
    this.dragBackground();
  }

  private getAllNodes() {
    return Object.entries(this.props.nodes).map(item => {
      return {
        uuid: item[0],
        data: item[1]
      };
    });
  }
  private findNodes(value: string) {
    const nodes = this.getAllNodes();
    return nodes.filter(item => {
      const nodeItem: any = item.data.node.actions;
      if (nodeItem.length > 0) {
        //find by type
        const foundTitle = i18n
          .t(`actions.${nodeItem[0].type}.name`)
          .toLocaleLowerCase()
          .includes(value.toLocaleLowerCase());
        //find by content
        const foundContent = nodeItem[0].hasOwnProperty('text')
          ? item.data.node.actions[0].text.includes(value)
          : nodeItem[0].hasOwnProperty('result_name')
          ? item.data.node.actions[0].result_name.includes(value)
          : false;

        return foundTitle || foundContent;
      } else {
        return false;
      }
    });
  }

  private dragBackground() {
    const canvasBg = document.getElementById('panzoom');
    if (this.props.search.nodes[this.props.search.selected] && canvasBg) {
      const uuid = this.props.search.nodes[this.props.search.selected].uuid;
      const ui = this.props.search.nodes[this.props.search.selected].data.ui.position;
      const width = window.innerWidth / 2;
      const height = window.innerHeight / 2;
      canvasBg.style.transform = `matrix(1, 0, 0, 1, ${width - ui.left}, ${height - ui.top})`;
      this.applyFilter(uuid);
    }
  }

  private applyFilter(uuid: string) {
    this.props.search.nodes.forEach(item => {
      const node = document.getElementById(item.uuid);
      if (node) {
        node.style.filter = 'none';
        if (item.uuid !== uuid && uuid !== 'remove') {
          node.style.filter = 'grayscale(1)';
        }
      }
    });
  }

  private toggleMoveSelected(type: 'up' | 'down') {
    switch (type) {
      case 'up':
        this.props.handleSearchChange({
          active: true,
          value: this.props.search.value,
          nodes: this.props.search.nodes,
          selected: this.props.search.selected - 1 < 0 ? 0 : this.props.search.selected - 1
        });
        break;
      case 'down':
        var down = this.props.search.selected - 1;
        if (down < 0) {
          down = 0;
        }
        this.props.handleSearchChange({
          active: true,
          value: this.props.search.value,
          nodes: this.props.search.nodes,
          selected:
            this.props.search.selected < this.props.search.nodes.length - 1
              ? this.props.search.selected + 1
              : this.props.search.nodes.length - 1
        });
        break;
    }
    this.dragBackground();
  }

  private closeSearch() {
    this.props.handleSearchChange({
      active: false,
      value: this.props.search.value,
      nodes: this.props.search.nodes,
      selected: 0
    });
    this.applyFilter('remove');
  }

  public render(): JSX.Element {
    return (
      <div className={styles.search_card}>
        <div className={styles.icon}>
          <UnnnicIcon icon="search-1" size="avatar-nano" scheme="neutral-dark" />
        </div>

        <div className={styles.input}>
          <TextInputElement
            name={''}
            placeholder={''}
            entry={{ value: this.props.search.value }}
            onChange={value => this.handleInput(value)}
            autocomplete={true}
          />
        </div>
        <div className={styles.buttons}>
          <div className={styles.button}>
            <Button
              name={''}
              onClick={() => this.toggleMoveSelected('down')}
              type={ButtonTypes.ghost}
              iconName="arrow-button-down-1"
              size="small"
            />
          </div>
          <div>
            <Button
              name={''}
              onClick={() => this.toggleMoveSelected('up')}
              type={ButtonTypes.ghost}
              iconName="arrow-button-up-1"
              size="small"
            />
          </div>
        </div>
        <span className={styles.number}>
          {this.props.search.selected + 1}/{this.props.search.nodes.length}
        </span>
        <Button
          name={''}
          onClick={() => this.closeSearch()}
          type={ButtonTypes.ghost}
          iconName="close-1"
          size="small"
          leftSpacing={true}
        />
      </div>
    );
  }
}

const mapStateToProps = ({ flowContext: { search, nodes } }: AppState) => {
  return {
    search,
    nodes
  };
};

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
  bindActionCreators({ handleSearchChange, onUpdateCanvasPositions }, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SearchBar);
