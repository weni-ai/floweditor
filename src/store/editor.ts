import { FlowPosition } from 'flowTypes';
import ActionTypes, { UpdateEditorState } from 'store/actionTypes';
import Constants from 'store/constants';
import { Asset, RenderNode } from 'store/flowContext';

// tslint:disable:no-shadowed-variable
export interface DragSelection {
  startX?: number;
  startY?: number;
  currentX?: number;
  currentY?: number;
}

export interface DebugState {
  showUUIDs: boolean;
}

export interface CanvasPositions {
  [uuid: string]: FlowPosition;
}

export interface Activity {
  nodes: { [uuid: string]: number };
  segments: { [exitToNodeKey: string]: number };
  recentMessages?: { [key: string]: RecentMessage[] };
  is_starting?: boolean;
}

export interface RecentMessage {
  sent: string;
  text: string;
}

export interface Warning {
  name: string;
  type: string;
}

export interface Warnings {
  [uuid: string]: Warning;
}

export interface EditorState {
  currentRevision: number | null;
  simulating: boolean;
  language: Asset | null;
  translating: boolean;
  fetchingFlow: boolean;
  ghostNode: RenderNode | null;
  dragActive: boolean;
  dragStartTime: number;
  dragDownPosition: FlowPosition | null;
  dragNodeUUID: string | null;
  dragGroup: boolean;
  dragSelection: DragSelection | null;
  selectionActive: boolean;
  debug?: DebugState | null;

  // which poptab is popped
  popped: string;

  warnings: Warnings;

  modalMessage?: ModalMessage;
  saving?: boolean;

  // the currently shown activity, can be
  // simulation or live
  activity: Activity;

  // the current live activity
  liveActivity: Activity;

  // interval in millis we should refresh activity
  activityInterval: number;

  // is our page visible or tabbed away
  visible: boolean;

  scrollToNode: string;
  scrollToAction: string;

  currentGuide: string;
  guidingStep: number;

  mouseState: MouseState;
}

export interface ModalMessage {
  title: string;
  body: string;
}

export enum MouseState {
  SELECT = 'select',
  DRAG = 'drag',
  DRAGGING = 'dragging'
}

export const EMPTY_DRAG_STATE: any = {
  dragStartTime: 0,
  dragDownPosition: null,
  dragActive: false,
  dragNodeUUID: null,
  dragGroup: false,
  dragSelection: null
};

// Initial state
export const initialState: EditorState = {
  currentRevision: null,
  simulating: false,
  translating: false,
  language: null,
  fetchingFlow: false,
  dragStartTime: 0,
  dragDownPosition: null,
  dragActive: false,
  dragNodeUUID: null,
  dragGroup: false,
  dragSelection: null,
  ghostNode: null,
  selectionActive: false,
  debug: null,
  warnings: {},

  popped: null,

  activity: { segments: {}, nodes: {} },
  liveActivity: { segments: {}, nodes: {} },
  activityInterval: 5000,
  visible: true,

  scrollToNode: null,
  scrollToAction: null,

  currentGuide: null,
  guidingStep: -1,

  mouseState: MouseState.SELECT
};

// Action Creator
export const updateEditorState = (editorState: EditorState): UpdateEditorState => ({
  type: Constants.UPDATE_EDITOR_STATE,
  payload: {
    editorState
  }
});

/* Reducer */
export const editorState = (editorState: EditorState = initialState, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_EDITOR_STATE:
      return action.payload!.editorState;
    default:
      return editorState;
  }
};

// export our reducer
export default editorState;
