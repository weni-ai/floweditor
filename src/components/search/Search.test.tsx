import React from 'react';
import { SearchBar, SearchStoreProps } from './SearchBar';
import { act, fireEvent, fireUnnnicInputChangeText, render } from 'test/utils';
import { NodeComp } from '../flow/node/Node';
import { MouseState } from '../../store/editor';
import { RenderNodeMap } from '../../store/flowContext';
import { shallowToJson } from 'enzyme-to-json';

const renderNodes = (nodes: RenderNodeMap) => {
  return Object.entries(nodes).map(([uuid, node]) => {
    return (
      <NodeComp
        key={uuid}
        nodeUUID={uuid}
        plumberMakeTarget={vi.fn()}
        plumberRecalculate={vi.fn()}
        plumberMakeSource={vi.fn()}
        plumberRemove={vi.fn()}
        plumberConnectExit={vi.fn()}
        plumberUpdateClass={vi.fn()}
        startingNode={false}
        onlyNode={false}
        selected={false}
        results={undefined}
        language={undefined}
        languages={undefined}
        activeCount={0}
        translating={false}
        simulating={false}
        debug={undefined}
        renderNode={node}
        issues={[]}
        onAddToNode={vi.fn()}
        onOpenNodeEditor={vi.fn()}
        removeNode={vi.fn()}
        mergeEditorState={vi.fn()}
        scrollToNode={''}
        scrollToAction={''}
        mouseState={MouseState.SELECT}
        brainInfo={undefined}
      />
    );
  });
}

// @ts-ignore
const nodes: RenderNodeMap = {
  '11f18b66-37ce-43f0-a74e-4638eb7e5921': {
    node: {
      uuid: '11f18b66-37ce-43f0-a74e-4638eb7e5921',
      actions: [
        {
          attachments: [],
          text: 'teste1',
          // @ts-ignore
          type: 'send_msg',
          quick_replies: [],
          uuid: 'bbb43f9d-d254-44ef-a9d1-7a3e392f972c',
        },
      ],
      exits: [
        {
          uuid: '07e2e3d2-d7cb-4905-93b8-50598ec4fdf3',
          destination_uuid: 'f5804ae2-2103-4b55-b17a-129b359d39a2',
        },
      ],
    },
    ui: {
      position: {
        left: 160,
        top: 100,
      },
      // @ts-ignore
      type: 'execute_actions',
    },
    inboundConnections: {},
  },
  'f5804ae2-2103-4b55-b17a-129b359d39a2': {
    node: {
      uuid: 'f5804ae2-2103-4b55-b17a-129b359d39a2',
      actions: [
        {
          attachments: [],
          text: 'teste2',
          // @ts-ignore
          type: 'send_msg',
          quick_replies: [],
          uuid: '77e4f770-42ad-4573-95cf-f7d96a020dd9',
        },
      ],
      exits: [
        {
          uuid: '72e44e5c-777e-41de-a35c-c52f418d6c6c',
          destination_uuid: null,
        },
      ],
    },
    ui: {
      position: {
        left: 129,
        top: 409,
      },
      // @ts-ignore
      type: 'execute_actions',
    },
    inboundConnections: {
      '07e2e3d2-d7cb-4905-93b8-50598ec4fdf3': '11f18b66-37ce-43f0-a74e-4638eb7e5921',
    },
  },
  '2eeba668-b742-4982-b8da-2cb3726903e5': {
    node: {
      uuid: '2eeba668-b742-4982-b8da-2cb3726903e5',
      actions: [
        {
          attachments: [],
          text: 'different',
          // @ts-ignore
          type: 'send_msg',
          quick_replies: [],
          uuid: '87e4f770-42ad-4573-95cf-f7d96a020dd9',
        },
      ],
      exits: [
        {
          uuid: '82e44e5c-777e-41de-a35c-c52f418d6c6c',
          destination_uuid: null,
        },
      ],
    },
    ui: {
      position: {
        left: 129,
        top: 409,
      },
      // @ts-ignore
      type: 'execute_actions',
    },
  },
};

const baseProps: SearchStoreProps = {
  search: {
    isSearchOpen: true,
    value: '',
    selected: 0,
    nodes: [],
  },
  nodes: nodes,
  handleSearchChange: vi.fn(),
};

describe(SearchBar.name, () => {
  
  vi.spyOn(document, 'createTreeWalker').mockImplementation((node: any) => {
    return {
      nextNode: vi.fn(),
    };
  })

  vi.spyOn(document, 'getElementById').mockImplementation((elementId: string) => {
    return {
      // @ts-expect-error Property 'text' does not exist on type 'Action'
      // eslint-disable-next-line prettier/prettier
      innerText: nodes[elementId]?.node.actions[0].text,
      querySelectorAll: vi.fn((): [] => {
        return []
      }),
    } as HTMLElement
  })

  beforeEach(() => {
    vi.clearAllMocks();
  })  

  it('should render', () => {
    const { baseElement } = render(
      <>
        <SearchBar {...baseProps} />
        {renderNodes(nodes)}
      </>
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should call handleSearchChange with correct nodes', async () => {
    const { getByTestId } = render(
      <>
        <SearchBar {...baseProps} />
        {renderNodes(nodes)}
      </>
    );
    
    expect(baseProps.handleSearchChange).not.toHaveBeenCalled();
    // change the search value
    const input = getByTestId('searchInput');
    fireUnnnicInputChangeText(input, 'test');

    expect(baseProps.handleSearchChange).toHaveBeenCalledTimes(1);
    expect(baseProps.handleSearchChange).toHaveBeenCalledWith({
      isSearchOpen: true,
      value: 'test',
      nodes: [
        {
          uuid: '11f18b66-37ce-43f0-a74e-4638eb7e5921',
          data: nodes['11f18b66-37ce-43f0-a74e-4638eb7e5921'],
        },
        { 
          uuid: 'f5804ae2-2103-4b55-b17a-129b359d39a2',
          data: nodes['f5804ae2-2103-4b55-b17a-129b359d39a2'],
        }
      ],
      selected: 0,
    });

    fireUnnnicInputChangeText(input, 'different');
    expect(baseProps.handleSearchChange).toHaveBeenCalledTimes(2);
    expect(baseProps.handleSearchChange).toHaveBeenCalledWith({
      isSearchOpen: true,
      value: 'different',
      nodes: [
        {
          uuid: '2eeba668-b742-4982-b8da-2cb3726903e5',
          data: nodes['2eeba668-b742-4982-b8da-2cb3726903e5'],
        }
      ],
      selected: 0,
    });
  })

  it('should iter through the search results', async () => {
    const { getByTestId, rerender } = render(
      <>
        <SearchBar {...baseProps} />
        {renderNodes(nodes)}
      </>
    );

    expect(baseProps.handleSearchChange).not.toHaveBeenCalled();
    // change the search value
    const input = getByTestId('searchInput');
    fireUnnnicInputChangeText(input, 'test');

    const searchMatchNodes = [
      {
        uuid: '11f18b66-37ce-43f0-a74e-4638eb7e5921',
        data: nodes['11f18b66-37ce-43f0-a74e-4638eb7e5921'],
      },
      { 
        uuid: 'f5804ae2-2103-4b55-b17a-129b359d39a2',
        data: nodes['f5804ae2-2103-4b55-b17a-129b359d39a2'],
      }
    ]

    rerender(
      <>
        <SearchBar {...baseProps} search={{...baseProps.search, value: 'test', nodes: searchMatchNodes}} />
        {renderNodes(nodes)}
      </>
    )

    expect(baseProps.handleSearchChange).toHaveBeenCalledWith({
      isSearchOpen: true,
      value: 'test',
      nodes: searchMatchNodes,
      selected: 0,
    });    

    const downButton = getByTestId('downButton');
    fireEvent.click(downButton);

    expect(baseProps.handleSearchChange).toHaveBeenCalledWith({
      isSearchOpen: true,
      value: 'test',
      nodes: searchMatchNodes,
      selected: 1,
    });

    // should go down on when enter is typed, also checks for circular navigation
    fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' });

    expect(baseProps.handleSearchChange).toHaveBeenCalledWith({
      isSearchOpen: true,
      value: 'test',
      nodes: searchMatchNodes,
      selected: 0,
    });

    // should go up
    const upButton = getByTestId('upButton');
    fireEvent.click(upButton);

    expect(baseProps.handleSearchChange).toHaveBeenCalledWith({
      isSearchOpen: true,
      value: 'test',
      nodes: searchMatchNodes,
      selected: 1,
    });
  })
});
