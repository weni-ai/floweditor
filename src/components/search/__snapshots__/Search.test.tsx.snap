// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render searchbar 1`] = `
<body>
  <div>
    <div
      class="search_card"
    >
      <div
        class="icon"
      >
        <div
          data-use-vue-component-wrap=""
          style="display: inline-block;"
        >
          <span
            class="unnnic-icon unnnic-icon__size--avatar-nano unnnic-icon-scheme--neutral-dark"
            data-passed-props="[object Object]"
            data-v-4ce83152=""
          >
            <svg
              fill="none"
              viewBox="0 0 40 40"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="primary"
                d="M34.0615 35.0005C33.8116 35.0005 33.5754 34.903 33.3992 34.7256L25.2575 26.5826C24.3065 27.3961 23.2355 28.0597 22.067 28.5559C20.5312 29.2082 18.9178 29.5394 17.272 29.5394C15.7011 29.5394 14.1578 29.2382 12.6819 28.6421C9.64389 27.4161 7.26574 25.0817 5.98481 22.0662C4.70388 19.052 4.67263 15.7178 5.89858 12.6811C7.12452 9.64433 9.45894 7.26492 12.4732 5.98399C14.009 5.33041 15.6211 5.00049 17.2682 5.00049C18.8391 5.00049 20.3825 5.30291 21.8583 5.89776C24.8963 7.12371 27.2745 9.45812 28.5554 12.4736C29.8363 15.4891 29.8676 18.822 28.6416 21.8588C28.1393 23.1035 27.4469 24.2444 26.5809 25.2554L34.7251 33.4009C35.09 33.7658 35.09 34.3606 34.7251 34.7268C34.5477 34.903 34.3115 35.0005 34.0615 35.0005ZM17.2682 6.87502C15.8748 6.87502 14.5077 7.15495 13.2068 7.70856C10.6524 8.79454 8.67413 10.809 7.63564 13.3821C6.59715 15.9552 6.6234 18.7783 7.70813 21.3327C8.7941 23.887 10.8086 25.8653 13.3817 26.9038C14.6326 27.4086 15.9411 27.6636 17.2707 27.6636C18.6641 27.6636 20.0313 27.3836 21.3322 26.83C22.5494 26.3127 23.6404 25.5828 24.5739 24.6606C24.5864 24.6431 24.6002 24.6281 24.6139 24.6143C24.6339 24.5943 24.6526 24.5781 24.6701 24.5656C25.6236 23.6021 26.381 22.4499 26.9021 21.1565C27.9406 18.5833 27.9143 15.7603 26.8283 13.2059C25.7436 10.6516 23.7279 8.67332 21.1548 7.63483C19.9063 7.13121 18.5979 6.87502 17.2682 6.87502Z"
                fill="#3B414D"
              />
            </svg>
          </span>
        </div>
      </div>
      <div
        class="input"
      >
        <div>
          <div
            class="completion_wrapper"
          >
            <div
              data-testid=""
            >
              <div
                data-use-vue-component-wrap=""
                style="all: unset;"
              >
                <div
                  class="unnnic-form sm"
                  data-passed-props="[object Object]"
                  data-v-48c0c241=""
                >
                  <!---->
                  <div
                    class="unnnic-form-input text-input size--sm"
                    data-v-400546b3=""
                    data-v-48c0c241=""
                    mask=""
                    v-bind="[object Object]"
                  >
                    <input
                      class="input-itself input size-sm normal"
                      data-passed-props="[object Object]"
                      data-v-400546b3=""
                      data-v-97b44e8c=""
                      placeholder=""
                      size="5"
                      type="text"
                      v-bind="[object Object]"
                    />
                    <!---->
                    <!---->
                  </div>
                  <!---->
                </div>
              </div>
            </div>
            <div
              class="options_wrapper"
              data-testid="temba_completion_options_"
              style="margin-top: 0px; width: 0px; display: none;"
            />
          </div>
        </div>
      </div>
      <div
        class="buttons"
      >
        <div
          class="button"
        >
          <div
            data-use-vue-component-wrap=""
          >
            <button
              class="unnnic-button-next unnnic-button-next--size-small unnnic-button-next--ghost unnnic-button-next-scheme--"
              data-passed-props="[object Object]"
              data-v-1eef067c=""
              style="margin-left: 0px; margin-top: 0px;"
            >
              <!---->
              <span
                class="unnnic-icon unnnic-icon__size--sm unnnic-icon-scheme--neutral-dark"
                data-v-1eef067c=""
                data-v-4ce83152=""
              >
                <svg
                  fill="none"
                  viewBox="0 0 40 40"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    class="primary"
                    d="M5.31471 14.0096L19.0432 29.5967C19.1676 29.7264 19.3169 29.8296 19.4822 29.9001C19.6474 29.9707 19.8253 30.007 20.005 30.007C20.1847 30.007 20.3625 29.9707 20.5278 29.9001C20.6931 29.8296 20.8424 29.7264 20.9668 29.5967L34.689 14.0096C34.7987 13.886 34.8825 13.7415 34.9353 13.5849C34.9882 13.4282 35.009 13.2625 34.9965 13.0977C34.984 12.9328 34.9385 12.7721 34.8627 12.6252C34.7869 12.4783 34.6824 12.3481 34.5553 12.2423L32.243 10.2963C31.9852 10.0789 31.6522 9.97163 31.316 9.99759C30.9797 10.0235 30.6671 10.1807 30.4457 10.435L20.2406 22.25C20.2112 22.2841 20.1749 22.3114 20.134 22.3302C20.0931 22.3489 20.0487 22.3587 20.0037 22.3587C19.9588 22.3587 19.9143 22.3489 19.8734 22.3302C19.8326 22.3114 19.7962 22.2841 19.7669 22.25L9.558 10.4338C9.33639 10.1806 9.02374 10.0252 8.68809 10.0013C8.52167 9.9865 8.35397 10.0052 8.19489 10.0563C8.03582 10.1074 7.8886 10.1898 7.76195 10.2988L5.44845 12.2423C5.32061 12.3479 5.21534 12.4781 5.13889 12.6252C5.06245 12.7723 5.0164 12.9332 5.0035 13.0985C4.99099 13.2635 5.01212 13.4294 5.06562 13.586C5.11911 13.7426 5.20386 13.8868 5.31471 14.0096Z"
                    fill="#3B414D"
                  />
                </svg>
              </span>
              <!---->
              <span
                class="unnnic-button-next__label"
                data-v-1eef067c=""
              >
                  
              </span>
              <!---->
            </button>
          </div>
        </div>
        <div>
          <div
            data-use-vue-component-wrap=""
          >
            <button
              class="unnnic-button-next unnnic-button-next--size-small unnnic-button-next--ghost unnnic-button-next-scheme--"
              data-passed-props="[object Object]"
              data-v-1eef067c=""
              style="margin-left: 0px; margin-top: 0px;"
            >
              <!---->
              <span
                class="unnnic-icon unnnic-icon__size--sm unnnic-icon-scheme--neutral-dark"
                data-v-1eef067c=""
                data-v-4ce83152=""
              >
                <svg
                  fill="none"
                  viewBox="0 0 40 40"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    class="primary"
                    d="M34.6851 25.9913L20.9567 10.4043C20.8323 10.2746 20.683 10.1714 20.5177 10.1009C20.3524 10.0304 20.1746 9.99402 19.9949 9.99402C19.8152 9.99402 19.6374 10.0304 19.4721 10.1009C19.3068 10.1714 19.1575 10.2746 19.0332 10.4043L5.311 25.9913C5.20125 26.1149 5.11748 26.2594 5.06465 26.416C5.01183 26.5727 4.99104 26.7384 5.00352 26.9032C5.01599 27.0681 5.06148 27.2288 5.13727 27.3757C5.21307 27.5226 5.31763 27.6528 5.44474 27.7586L7.75697 29.7046C8.01481 29.922 8.34778 30.0292 8.68402 30.0033C9.02027 29.9773 9.33283 29.8202 9.55426 29.5658L19.7593 17.751C19.7886 17.7169 19.825 17.6895 19.8659 17.6708C19.9067 17.652 19.9512 17.6423 19.9962 17.6423C20.0411 17.6423 20.0856 17.652 20.1265 17.6708C20.1673 17.6895 20.2037 17.7169 20.233 17.751L30.4418 29.5621C30.5518 29.6887 30.6857 29.7924 30.8358 29.8673C30.986 29.9421 31.1494 29.9867 31.3167 29.9983C31.4831 30.0131 31.6508 29.9944 31.8099 29.9434C31.969 29.8923 32.1162 29.8098 32.2429 29.7008L34.5563 27.7536C34.6832 27.6476 34.7874 27.5174 34.863 27.3704C34.9386 27.2235 34.984 27.0629 34.9965 26.8982C35.0089 26.7334 34.9883 26.5678 34.9356 26.4112C34.883 26.2546 34.7995 26.1101 34.6901 25.9863L34.6851 25.9913Z"
                    fill="#3B414D"
                  />
                </svg>
              </span>
              <!---->
              <span
                class="unnnic-button-next__label"
                data-v-1eef067c=""
              >
                  
              </span>
              <!---->
            </button>
          </div>
        </div>
      </div>
      <span
        class="number"
      >
        1
        /
        1
      </span>
      <div
        data-use-vue-component-wrap=""
      >
        <button
          class="unnnic-button-next unnnic-button-next--size-small unnnic-button-next--ghost unnnic-button-next-scheme--"
          data-passed-props="[object Object]"
          data-v-1eef067c=""
          style="margin-top: 0px;"
        >
          <!---->
          <span
            class="unnnic-icon unnnic-icon__size--sm unnnic-icon-scheme--neutral-dark"
            data-v-1eef067c=""
            data-v-4ce83152=""
          >
            <svg
              fill="none"
              viewBox="0 0 40 40"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="primary"
                d="M34.0592 34.9982C33.9361 34.9985 33.8142 34.9744 33.7004 34.9272C33.5867 34.88 33.4835 34.8107 33.3969 34.7232L19.9995 21.3271L6.60213 34.7232C6.51561 34.811 6.4125 34.8808 6.29879 34.9284C6.18507 34.976 6.06303 35.0005 5.93976 35.0005C5.81648 35.0005 5.69444 34.976 5.58073 34.9284C5.46701 34.8808 5.3639 34.811 5.27739 34.7232C5.10209 34.5471 5.00367 34.3087 5.00367 34.0602C5.00367 33.8117 5.10209 33.5734 5.27739 33.3972L18.6748 19.9999L5.27739 6.60374C5.09992 6.42753 5.00244 6.19132 5.00244 5.94137C5.00244 5.69142 5.09992 5.45522 5.27739 5.279C5.4536 5.10154 5.68981 5.00406 5.93976 5.00406C6.18971 5.00406 6.42591 5.10154 6.60213 5.279L19.9995 18.6751L33.3969 5.27775C33.4834 5.18994 33.5865 5.1202 33.7002 5.0726C33.8139 5.025 33.936 5.00049 34.0592 5.00049C34.1825 5.00049 34.3046 5.025 34.4183 5.0726C34.532 5.1202 34.6351 5.18994 34.7216 5.27775C34.8092 5.36454 34.8787 5.46782 34.9261 5.58161C34.9735 5.69541 34.9979 5.81747 34.9979 5.94075C34.9979 6.06403 34.9735 6.18609 34.9261 6.29988C34.8787 6.41367 34.8092 6.51695 34.7216 6.60374L21.3242 20.0011L34.7216 33.3985C34.8969 33.5746 34.9953 33.813 34.9953 34.0615C34.9953 34.31 34.8969 34.5484 34.7216 34.7245C34.6347 34.8115 34.5315 34.8805 34.4178 34.9275C34.3041 34.9745 34.1822 34.9985 34.0592 34.9982Z"
                fill="#3B414D"
              />
            </svg>
          </span>
          <!---->
          <span
            class="unnnic-button-next__label"
            data-v-1eef067c=""
          >
              
          </span>
          <!---->
        </button>
      </div>
    </div>
  </div>
</body>
`;
